PowerEditor/src/WinControls/ProjectPanel/ProjectPanel.cpp

```diff
@@ -534,6 +743,259 @@ void ProjectPanel::openSelectFile()
+		if( result.empty() ) // drive
+		{
+			result = filePath;
+			result.erase(lastSlashIdx);
+		}
+	}
+
+	if (!filters.empty())
+	{
+		result += TEXT(" (");
+		result += stringJoin(filters, TEXT(";"), true);
+		result += TEXT(")");
+	}
+
+	return result;
+
+}
+
+void ProjectPanel::itemVisibilityChanges(HTREEITEM hItem, bool visible)
+{
+	TVITEM tvItem;
+	tvItem.mask = TVIF_PARAM;
+	tvItem.hItem = hItem;
+	::SendMessage(_treeView.getHSelf(), TVM_GETITEM, 0, reinterpret_cast<LPARAM>(&tvItem));
+	ProjectPanelData& data = *reinterpret_cast<ProjectPanelData*>(tvItem.lParam);
+	bool expanded = (tvItem.state & TVIS_EXPANDED) != 0;
+	if (data.isDirectory())
+	{
+		if (visible && expanded)
+			data.watchDir(true);
+		if (!visible || !expanded)
+			data.watchDir(false);
+	}
+
+	for (HTREEITEM hChildItem = _treeView.getChildFrom(hItem);
+    hChildItem != NULL;
+    hChildItem = _treeView.getNextSibling(hChildItem))
+    {
+		itemVisibilityChanges(hChildItem, visible);
+	}
+
+}
+
+void ProjectPanel::expandOrCollapseDirectory(bool expand, HTREEITEM hItem)
+{
+
+	TVITEM tvItem;
+	tvItem.mask = TVIF_PARAM;
+	tvItem.hItem = hItem;
+	::SendMessage(_treeView.getHSelf(), TVM_GETITEM, 0, reinterpret_cast<LPARAM>(&tvItem));
+	ProjectPanelData& data = *reinterpret_cast<ProjectPanelData*>(tvItem.lParam);
+	if (!(data.isDirectory() || data.isBaseDirectory()) )
+		return;
+
+	if (expand)
+	{
+		removeDummies(hItem);
+		if (data.isDirectory())
+			data.watchDir(true);
+	}
+	else
+	{
+		if (data.isDirectory())
+			data.watchDir(false);
+	}
+
+	for (HTREEITEM hChildItem = _treeView.getChildFrom(hItem);
+    hChildItem != NULL;
+    hChildItem = _treeView.getNextSibling(hChildItem))
+    {
+		itemVisibilityChanges(hChildItem, expand);
+	}
+
+	_directoryWatcher.update();
+
+}
+
+bool ProjectPanel::setFilters(const std::vector<generic_string>& filters, HTREEITEM hItem)
+{
+	TVITEM tvItem;
+	tvItem.mask = TVIF_PARAM;
+	tvItem.hItem = hItem;
+	::SendMessage(_treeView.getHSelf(), TVM_GETITEM, 0, reinterpret_cast<LPARAM>(&tvItem));
+	ProjectPanelData& data = *reinterpret_cast<ProjectPanelData*>(tvItem.lParam);
+
+	if (data.isDirectory() || data.isBaseDirectory())
+	{
+		if (data.isBaseDirectory() && data._filters == filters)
+			return false;
+
+		data._filters = filters;
+	}
+
+	if (data.isWatching())
+	{
+		data.watchDir(true);
+	}
+
+	for (HTREEITEM hItemNode = _treeView.getChildFrom(hItem);
+		hItemNode != NULL;
+		hItemNode = _treeView.getNextSibling(hItemNode))
+	{
+		setFilters(filters,hItemNode);
+	}
+
+	updateBaseDirName(hItem);
+	return true;
+}
+
+void ProjectPanel::updateBaseDirName(HTREEITEM hItem)
+{
+	TCHAR textBuffer[MAX_PATH];
+	TVITEM tvItem;
+	tvItem.mask = TVIF_TEXT | TVIF_PARAM;
+	tvItem.pszText = textBuffer;
+	tvItem.cchTextMax = MAX_PATH;
+	tvItem.hItem = hItem;
+	::SendMessage(_treeView.getHSelf(), TVM_GETITEM, 0, reinterpret_cast<LPARAM>(&tvItem));
+	ProjectPanelData& data = *reinterpret_cast<ProjectPanelData*>(tvItem.lParam);

```

Comment:
You ***absolutely*** should check to see if `::SendMessage` returned `FALSE` before using `tvItem.lParam`.